swift/ql/test/library-tests/dataflow/dataflow/test.swift

```diff
@@ -129,7 +129,7 @@ func forwarder() {
         (i: Int) -> Int in
         return 0
     })
-    sink(arg: clean)
+    sink(arg: clean) // $ SPURIOUS: flow=128
```

Comment:
Yes. We're missing call contexts for the call above:
```cpp
forward(arg: source(), lambda: {
  (i: Int) -> Int in
    return 0
})
```
so we actually get flow like:
```
source() (i.e., the argument) -> arg (i.e., the parameter of `forward`) -> arg (i.e., the argument of lambda(arg)) -> i (the parameter of on line 123) -> i (the return on line 124) -> clean
```

There can be two explanations of this:
- Swift doesn't implement dispatch through call contexts (see `mayBenefitFromCallContext` and `viableImplInCallContext` in `DataFlowDispatch`).
- The call contexts exists, but is being thrown away because there's an issue with an enclosing callable. I actually believe this is the case based on @aschackmull's feedback [here](https://github.com/github/codeql/pull/14577#discussion_r1371503179).

I'll look into this as well.

Reply:
This was fixed in 2465cc20f0108a9b80c85e438695cb1f43b2f3eb (the answer was indeed a version of case 2).

Reply:
We've lost two good results in that change - and they look like fairly straightforward ones, unless I'm missing something?

Reply:
We're losing this one:
```swift
func setEscape() {
  let x = source("setEscape", 0)
  escape = {
    sink(x) // $ MISSING: hasValueFlow=setEscape
    return x + 1
  }
}
```
because we don't model global flow in Swift. So we don't manage to flow from the definition of `escape` to the place where it's actually called:
```swift
func callEscape() {
  setEscape()
  sink(escape?()) // $ MISSING: hasTaintFlow=setEscape
}
```
In an earlier commit we accidentially got this flow because we flowed from `let x = source("setEscape", 0)` and straight into the closure (which violates all kinds of dataflow invariants), but when I fixed that in 2465cc20f0108a9b80c85e438695cb1f43b2f3eb the result disappears again, and we're only gonna get it back if we implement global flow in Swift.

For this one:
```swift
func captureList() {
  let y: Int = source("captureList", 123);
  { [x = hello()] () in
     sink(x) // $ MISSING: hasValueFlow=hello
     sink(y) // $ MISSING: hasValueFlow=captureList
  }()
}
```
I feel like we should be able to catch `sink(y)` (but for some reason we still don't). We don't catch the `sink(x)` one because we're missing an SSA definition for `x = hello()` and that work is blocked on https://github.com/github/codeql/pull/14567 (which I've promised Paolo to look at after this is merged).

Reply:
Yep I'm comfortable calling the the above an issue with capture lists. I've added a version of this test in 65e13aa5edf5a8ff3446e41c0aab3f4ed2ce3266 without a capture list, and that works fine. So I think we should delay further work on this until https://github.com/github/codeql/pull/14567 is merged.